File: distros/T/Text-ClearSilver/example/dbi.pl
#!perl -w
use 5.010_000;
use strict;
use Config;
use DBI;
use Text::ClearSilver;

my $dbh = DBI->connect('dbi:SQLite:dbname=:memory:', undef, undef,
    { PrintError => 0, RaiseError => 1} );

$dbh->do(<<'SQL');
create table perl_config (
    name  text primary key,
    value text
);
SQL

my $sth = $dbh->prepare(<<'SQL');
insert into perl_config values(?, ?);
SQL

while(my($k, $v) = each %Config){
    $sth->execute($k, $v);
}

# output

$sth = $dbh->prepare(<<'SQL');
select * from perl_config where name like ?;
SQL

#use Data::Dumper; print Dumper tied(%$sth);
$sth->execute($ARGV[0] || '%');

my %row;
$sth->bind_columns(\@row{ @{$sth->{NAME_lc}} });

my $hdf = Text::ClearSilver::HDF->new;
my $i = 0;
while($sth->fetch) {
    $hdf->set_value("PerlConfig.$i.name"  => $row{name});
    $hdf->set_value("PerlConfig.$i.value" => $row{value} // 'undef');
    $i++;
}

#print $hdf->dump;

my $tcs = Text::ClearSilver->new();

$tcs->process(\<<'TCS', $hdf);
<?cs each:item = PerlConfig ?><?cs var:item.name ?> = <?cs var:item.value ?>
<?cs /each ?>
TCS

File: distros/R/Ruby/example/dbi.rb
#!perl -w
use DBI;
use Config qw(%Config);
$| = 1; # autoflush
use Ruby::Run;
puts "DBI demo";

pcfg = Perl["%Config"];

dbh = Perl["DBI"].connect("dbi:DBM:", Perl::undef, Perl::undef,
	{ "AutoCommit" => 0, "RaiseError" => 1}.to_perl());

puts "Initializing ...";

dbh.do("CREATE TABLE test (name VARCHAR PRIMARY KEY, value VARCHAR NOT NULL)");

sth = dbh.prepare("INSERT INTO test VALUES(?,?)");

pcfg.each_pair {|key,value|
	sth.execute(key, value);
}


sth = dbh.prepare("SELECT * FROM test WHERE name LIKE ?".to_perl());

puts "Selecting from Perl::Config,",
		"please input a key of the config hash (the wildcard '*' is available).",
		"CTRL-D to exit from this program.";

print "> ";

STDIN.each do |line|
	sth.execute(line.gsub(/\*/, '%').chomp());

	while( (row = sth.fetchrow_hashref()).defined? )
		puts "%s=%s" % [ row["name"], row["value"] ];
	end

	print "> ";
end

puts "\nCreaning up ...";

dbh.do("DROP TABLE test");

